R/data_separate.R

```diff
@@ -0,0 +1,403 @@
+                          append = FALSE,
+                          ignore_case = FALSE,
+                          verbose = TRUE,
+                          regex = FALSE,
+                          ...) {
+  # we need at least one explicit choice for either `new_columns` or `guess_columns`
+  if (is.null(new_columns) && is.null(guess_columns)) {
+    insight::format_error("Cannot separate values. Either `new_columns` or `guess_columns` must be provided.")
+  }
+  # in case user did not provide names of new columns, we can try
+  # to guess number of columns per variable
+  guess_columns <- match.arg(guess_columns, choices = c("min", "max", "mode"))
+
+  # make sure we have valid options for fill and extra
+  fill <- match.arg(fill, choices = c("left", "right", "value_left", "value_right"))
+  extra <- match.arg(extra, choices = c("drop_left", "drop_right", "merge_left", "merge_right"))
+
+  # evaluate select/exclude, may be select-helpers
+  select <- .select_nse(select,
+    data,
+    exclude,
+    ignore_case,
+    regex = regex,
+    verbose = verbose
+  )
+
+  # make new_columns as list, this works with single and multiple columns
+  if (!is.null(new_columns) && !is.list(new_columns)) {
+    new_columns <- rep(list(new_columns), times = length(select))
+    # if we have multiple columns that were separated, we avoid duplicated
+    # column names of created variables by appending name of original column
+    # however, we don't have duplicated column names when we merge them together
+    # so don't create new column names when "merge_multiple" is FALSE.
+    make_unique_colnames <- length(select) > 1 && !merge_multiple
+  } else {
+    # we don't want to create own unique column names when user explicitly
+    # provided column names as a list, i.e. column names for each separated
+    # variable
+    make_unique_colnames <- FALSE
+  }
+
+  # make sure list of new column names is named
+  if (!is.null(new_columns) && is.null(names(new_columns))) {
+    names(new_columns) <- select
+  }
+
+  # iterate columns that should be split
+  split_data <- lapply(select, function(sep_column) {
+
+    # do we have known number of columns?
+    if (is.null(new_columns)) {
+      n_columns <- NULL
+    } else {
+      n_columns <- length(new_columns[[sep_column]])
+    }
+
+    # make sure we have a character that we can split
+    x <- data[[sep_column]]
+    if (!is.character(x)) {
+      x <- as.character(x)
+    }
+
+    # separate column into multiple strings
+    if (is.numeric(separator)) {
+      maxlen <- max(nchar(x), na.rm = TRUE)
+      starts <- c(0, separator)
+      ends <- c(separator - 1, maxlen)
+      separated_columns <- lapply(seq_along(starts), function(i) {
+        substr(x, starts[i], ends[i])
+      })
+      separated_columns <- as.data.frame(
+        t(do.call(cbind, separated_columns)),
+        stringsAsFactors = FALSE
+      )
+    } else {
+      separated_columns <- strsplit(x, separator)
```

Comment:
```suggestion
      separated_columns <- strsplit(x, separator, perl = TRUE)
```
I don't know if adding this has downsides but it is faster:
``` r
test <- rep("a.b.c.d", 1e6)

bench::mark(
  strsplit(test, "\\."),
  strsplit(test, "\\.", perl = TRUE),
  iterations = 10
)
#> # A tibble: 2 × 6
#>   expression                            min  median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>                       <bch:tm> <bch:t>     <dbl> <bch:byt>    <dbl>
#> 1 "strsplit(test, \"\\\\.\")"         1.36s    1.5s     0.686    7.63MB     1.60
#> 2 "strsplit(test, \"\\\\.\", perl… 586.81ms 593.6ms     1.69     7.63MB     3.94
```